dev/benchmarks/microbenchmarks/lib/geometry/matrix_utils_transform_rect_bench.dart

```diff
@@ -0,0 +1,139 @@
+// Copyright 2019 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math' as math;
+
+import 'package:flutter/painting.dart';
+import 'package:flutter/rendering.dart';
+import 'package:vector_math/vector_math_64.dart';
+
+import '../common.dart';
+
+const int _kNumIterations = 1000000;
+const int _kNumWarmUp = 10000;
+
+void main() {
+  assert(false, "Don't run benchmarks in checked mode! Use 'flutter run --release'.");
+  print('MatrixUtils.transformRect benchmark...');
+
+  Matrix4 _makePerspective(double radius, double angle, double perspective) {
+    return MatrixUtils.createCylindricalProjectionTransform(
+      radius: radius,
+      angle: angle,
+      perspective: perspective,
+    );
+  }
+
+  // The number of transforms (and perspective transforms) provided as data
+  // should be relatively prime to the number of rectangles or points provided
+  // so that the benchmark loops end up applying each transform to each point
+  // or rect. Benchmarks don't have asserts enabled, but essentially:
+  //  assert(_rects.length.gcd(_transforms.length) == 1);
+  //  assert(_rects.length.gcd(_perspectiveTransforms.length) == 1);
+  //  assert(_offsets.length.gcd(_perspectiveTransforms.length) == 1);
+  //  assert(_offsets.length.gcd(_transforms.length) == 1);
```

Comment:
Shall we run the above checks and throw an exception if they failed even in the release mode (as they don't count in the time measurements)? I believe that an exception or a failed test would guard these assumptions much better than a comment :-)

Alternatively, it might be convenient to drop the `gcd == 1` assumption by running `_kNumIterations` iterations per pair. For example:
```
  for (int i = 0; i < _kNumIterations; i += 1) {
    for (Rect rect in _rects) {
      for (transform in _perspectiveTransforms) {
        MatrixUtils.transformRect(transform, rect);
      }
    }
  }
```
That would make it easier to add new transforms/rects/offsets to the benchmark.

If a relatively constant running time is a concern, maybe we can compute `numIterations = _kNumIterations ~/ numPairs`. (In the example above, `numPairs = _rects.length * _perspectiveTransforms.length`.) And if `watch.elapsed` has a low overhead, we can do something like `while (watch.elapsed < kBenchmarkTime) { ... }`.

Reply:
I thought about trying to rescue the asserts or convert them to if statements as Dan suggested, but I decided it wasn't worth it in the long run. The validity of the benchmark isn't hurt if the conditions don't hold, but it does make the benchmark (ever so) slightly better.

The variation you give is interesting, but I wonder how much overhead is created compared to just using integer modulus. The benchmark is already very fast even with me bumping it to 1m iterations (and even faster when I get my fix in).

If I go with your double (triple) loop, I would probably use a regular for loop and indexing just to make sure that iterator creation doesn't create runtime garbage (since we are measuring an issue that was caused by excessive unnecessary memory allocation).

In the end, though, I don't think it is a critical issue to worry about (because of first paragraph above). Thoughts?

Reply:
Another issue I was trying to avoid was not having similar data between runs of the benchmark. Note that in Java with HotSpot, you could never be sure if the runtime compiler wouldn't detect that you were doing the same operation over and over again and optimize it out. Especially if you are running the same loop a million times - that would ramp up its optimizer and pretty soon it would be doing global analysis of whether the methods had side effects and whether you stored any results. We occasionally developed benchmarks whose performance suddenly went through the roof when the HotSpot team got more clever about inlines and results analysis.

Because of that, my benchmark philosophy has always been to make it as opaque to the compiler as I could that I was doing the same work over and over.  In this case, and because I reduced the number of data points to avoid benchmarking reading a large list of input data, I wanted to make it as obscure as possible that the operations were repeating.  The modulus relationship does that, but 3 nested loops spell it out more in language that the compiler might eventually be able to capitalize on and trivialize the benchmark.

Perhaps Dart doesn't have optimizations like that - for now? But perhaps it might get optimizations like that in the future.

Reply:
Cool. Can you please add your concern to the code as comments so someone new to the code could easily find this information?

As for `gcd != 1`, does it make the benchmarks only measure a limited combinations? For example, if `rects` and `transforms` both have `length = 5`, then we'll only benchmark 5 pairs, instead of 25?

Reply:
I resolved this by removing the concern.